apps/src/client.rs

```diff
@@ -600,3 +591,75 @@ pub fn connect(
 
     Ok(())
 }
+
+fn create_sockets(
+    poll: &mut mio::Poll, peer_addr: &std::net::SocketAddr, args: &ClientArgs,
+) -> (
+    Slab<mio::net::UdpSocket>,
+    HashMap<std::net::SocketAddr, usize>,
+    std::net::SocketAddr,
+) {
+    let mut sockets = Slab::with_capacity(std::cmp::max(args.addrs.len(), 1));
+    let mut src_addrs = HashMap::new();
+    let mut first_local_addr = None;
+
+    // Create UDP sockets backing the QUIC connection, and register them with
+    // the event loop. Check first user-provided addresses and keep the ones
+    // compatible with the address family of the peer.
+    for src_addr in args.addrs.iter().filter(|sa| {
+        (sa.is_ipv4() && peer_addr.is_ipv4()) ||
+            (sa.is_ipv6() && peer_addr.is_ipv6())
+    }) {
+        let socket = mio::net::UdpSocket::bind(*src_addr).unwrap();
+        let local_addr = socket.local_addr().unwrap();
+        let token = sockets.insert(socket);
+        src_addrs.insert(local_addr, token);
+        poll.registry()
+            .register(
+                &mut sockets[token],
+                mio::Token(token),
+                mio::Interest::READABLE,
+            )
+            .unwrap();
+        if first_local_addr.is_none() {
+            first_local_addr = Some(local_addr);
+        }
+    }
+
+    // If there is no such address, rely on the default INADDR_IN or IN6ADDR_ANY
+    // depending on the IP family of the server address. This is needed on macOS
+    // and BSD variants that don't support binding to IN6ADDR_ANY for both v4
+    // and v6.
+    if first_local_addr.is_none() {
+        let bind_addr = match peer_addr {
+            std::net::SocketAddr::V4(_) => "0.0.0.0:0",
+            std::net::SocketAddr::V6(_) => "[::]:0",
+        };
+        let bind_addr = bind_addr.parse().unwrap();
+        let socket = mio::net::UdpSocket::bind(bind_addr).unwrap();
+        let local_addr = socket.local_addr().unwrap();
+        let token = sockets.insert(socket);
+        src_addrs.insert(local_addr, token);
+        poll.registry()
+            .register(
+                &mut sockets[token],
+                mio::Token(token),
+                mio::Interest::READABLE,
+            )
+            .unwrap();
+        first_local_addr = Some(local_addr)
+    }
+
+    (sockets, src_addrs, first_local_addr.unwrap())
+}
+
+/// Generate a ordered list of 4-tuples on which the host should send packets,
+/// following a lowest-latency scheduling.
+fn lowest_latency_scheduler(
+    conn: &quiche::Connection,
+) -> impl Iterator<Item = (std::net::SocketAddr, std::net::SocketAddr)> {
+    use itertools::Itertools;
+    conn.path_stats()
+        .sorted_by_key(|p| p.rtt)
+        .map(|p| (p.local_addr, p.peer_addr))
```

Comment:
When there is a closed path, I'd like to add something like `filter(|p| p.active)` here, but that will break probing of new paths.
It seems not enough information is marked as public in order to implement packet scheduler outside of the quiche library.

Reply:
Yeah, this is likely that the current content of `PathStats` might not be sufficient to address all the possible scheduling decisions. Might be useful to list all the fields of interest, but not sure if this PR will have all of them.

Reply:
I guess I will at least need the following things, but I can locally patch the code for now and am not in a hurry.
The list would change as my project evolves, anyway. And I'm not sure `PathStats` is the right interface for all these things.

- a list of probing paths
- a list of ack eliciting paths
- "instantaneous" loss counter such as pto_count (for scheduling and path sanity checks, see the previous comment)

Reply:
@toshiiw I think if we want to stay in the current API philosophy of quiche, `PathStats` is fine.

Regarding the list of probing paths, this is already accessible since `PathStats` exposes the `PathValidationState`, so paths can be filtered through that field.

For the remaining points, it is still possible to extend the `PathStats` structure to also give access to, e.g., the `pto_count` (we already give access to some `Recovery` fields, so adding one is quite easy). For the ack eliciting, are you interested by paths requiring sending ack eliciting packets, or by paths having sent ack eliciting packets?

Reply:
Or, `send_on_path` can have a `hint` bool argument (or another function wrapper) to notify if the function is allowed to send on a different path than specified.